### ISSUE TITLE
Strange behavior on worker shutdown

### ISSUE THREAD

**Issue by Wonshtrum** [sozu-proxy/sozu#810]
Sōzu has two possible commands to shutdown a worker: SoftStop and HardStop. Both present some weird behavior.

## Confusing logs
Both commands trigger the `notify` method:
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/server.rs#L609-L620
They trickle down to `notify_proxys`, and call `handle_order`:
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/server.rs#L900-L904
and fall in the default case:
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/command/src/state.rs#L321-L328
which respectively logs:
```
SOZU ERROR    state cannot handle order message: SoftStop
SOZU ERROR    state cannot handle order message: HardStop
```

## Panic
Furthermore, for the `SoftStop`, the `shutting_down` member of the `Server` is given a value:
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/server.rs#L617
which, on following iterations, will trigger:
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/server.rs#L739-L743
Note here that the `SessionManager` is borrowed mutably.

Now, a session could be TCP, HTTP, or HTTPS (rustls/openssl). The TCP route is straightforward, `shutting_down` calls `close`
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/tcp.rs#L1004-L1006
`close` calls `close_backend`
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/tcp.rs#L923-L936
and in case the session was connected to a backend, it borrows mutably *again* the `SessionManager`:
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/tcp.rs#L789-L801
which crashes the worker `panicked at 'already borrowed: BorrowMutError'`.

The difference between TCP and HTTP(S) sessions is well explained in this comment:
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/lib.rs#L264-L268

The HTTP and two HTTPS sessions share the same logic so I will only follow the HTTP case:
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/http.rs#L1316-L1325
In case the session calls `close`, the same double mutable borrow occurs as for the TCP session, and the worker crashes.

## HTTP State doesn't shutdown
The 3 HTTP(S) sessions internally call `shutting_down` on their "Http State" (if their `protocol` is `Http`):
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/protocol/http/mod.rs#L517-L531
In case `front_buf` or `back_buf` are `None`, the condition evaluates to `false`. I don't think this should be the case. I think `None` buffers should be treated as empty buffers. For example, the current behavior prevents the closing of new sessions which don't have received anything yet since `Http::new` initializes `front_buf` and `back_buf` to `None`:
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/protocol/http/mod.rs#L152-L153
It also prevents the closing of sessions which received requests, responded, and have reset, since `reset` set `front_buf` back to `None`:
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/protocol/http/mod.rs#L199-L219
In my tests, `SoftStop` tries multiple times to stop sessions that are not transferring any more data (unsuccessfully). But ends up shutting down the worker once those inactive sessions timeout and are deleted.

## Summary
To sum up: shutting a worker down produces an ambiguous log, if it is a `SoftStop` the worker struggles to close sessions it should be able to close, and finally, if it can close a session it crashes by mutably borrowing the `SessionManager` two times.

**Comment by Keksoj** [comment]
This is excellent analysis. This issue to be tackled in the 0.14 milestone in my opinion.

**Comment by Wonshtrum** [comment]
While testing, I came across another strange behavior.
Suppose a TCP connection is established by a client and Sōzu is waiting for the "front RW readiness".
In the meantime, a `SoftStop` is sent to the worker. As said earlier it trickles down in `notify_proxys` which is dispatched on every proxy (TCP, HTTP, HTTPS).
Specifically, for the TCP proxy:
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/tcp.rs#L1283-L1293
which drains the listeners. If, in the same loop:
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/server.rs#L554
in following iterations, the "front RW readiness" is processed (meaning the `SoftStop` and front readiness are processed in this order in the same loop), the `ready` method is called:
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/server.rs#L651
and is passed down to the `ready` method of the corresponding session:
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/server.rs#L1596-L1597
which calls `ready_inner`:
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/tcp.rs#L991-L993
in our case, the session is still not connected to a backend, so it calls `connect_to_backend`:
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/tcp.rs#L623-L624
which tries to get the listener on which the request was sent:
https://github.com/sozu-proxy/sozu/blob/e3e4e2374b9315c60674278015a992e4394103d3/lib/src/tcp.rs#L831-L844
but in this (really rare corner case), the listeners have all been drained, so it panics:
```
panicked at 'no entry found for key', lib/src/tcp.rs:844:12
```

**Comment by FlorentinDUBOIS** [comment]
This issue is closed thanks to the pr #812
